File: paimon-core/src/main/java/org/apache/paimon/mergetree/compact/MergeTreeCompactTask.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.mergetree.compact;

import org.apache.paimon.compact.CompactDeletionFile;
import org.apache.paimon.compact.CompactResult;
import org.apache.paimon.compact.CompactTask;
import org.apache.paimon.compact.CompactUnit;
import org.apache.paimon.data.InternalRow;
import org.apache.paimon.io.DataFileMeta;
import org.apache.paimon.mergetree.SortedRun;
import org.apache.paimon.operation.metrics.CompactionMetrics;

import javax.annotation.Nullable;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.function.Supplier;

import static java.util.Collections.singletonList;

/** Compact task for merge tree compaction. */
public class MergeTreeCompactTask extends CompactTask {

    private final long minFileSize;
    private final CompactRewriter rewriter;
    private final int outputLevel;
    private final Supplier<CompactDeletionFile> compactDfSupplier;

    private final List<List<SortedRun>> partitioned;

    private final boolean dropDelete;
    private final int maxLevel;

    // metric
    private int upgradeFilesNum;

    public MergeTreeCompactTask(
            Comparator<InternalRow> keyComparator,
            long minFileSize,
            CompactRewriter rewriter,
            CompactUnit unit,
            boolean dropDelete,
            int maxLevel,
            @Nullable CompactionMetrics.Reporter metricsReporter,
            Supplier<CompactDeletionFile> compactDfSupplier) {
        super(metricsReporter);
        this.minFileSize = minFileSize;
        this.rewriter = rewriter;
        this.outputLevel = unit.outputLevel();
        this.compactDfSupplier = compactDfSupplier;
        this.partitioned = new IntervalPartition(unit.files(), keyComparator).partition();
        this.dropDelete = dropDelete;
        this.maxLevel = maxLevel;

        this.upgradeFilesNum = 0;
    }

    @Override
    protected CompactResult doCompact() throws Exception {
        List<List<SortedRun>> candidate = new ArrayList<>();
        CompactResult result = new CompactResult();

        // Checking the order and compacting adjacent and contiguous files
        // Note: can't skip an intermediate file to compact, this will destroy the overall
        // orderliness
        for (List<SortedRun> section : partitioned) {
            if (section.size() > 1) {
                candidate.add(section);
            } else {
                SortedRun run = section.get(0);
                // No overlapping:
                // We can just upgrade the large file and just change the level instead of
                // rewriting it
                // But for small files, we will try to compact it
                for (DataFileMeta file : run.files()) {
                    if (file.fileSize() < minFileSize) {
                        // Smaller files are rewritten along with the previous files
                        candidate.add(singletonList(SortedRun.fromSingle(file)));
                    } else {
                        // Large file appear, rewrite previous and upgrade it
                        rewrite(candidate, result);
                        upgrade(file, result);
                    }
                }
            }
        }
        rewrite(candidate, result);
        result.setDeletionFile(compactDfSupplier.get());
        return result;
    }

    @Override
    protected String logMetric(
            long startMillis, List<DataFileMeta> compactBefore, List<DataFileMeta> compactAfter) {
        return String.format(
                "%s, upgrade file num = %d",
                super.logMetric(startMillis, compactBefore, compactAfter), upgradeFilesNum);
    }

    private void upgrade(DataFileMeta file, CompactResult toUpdate) throws Exception {
        if (file.level() == outputLevel) {
            return;
        }

        if (outputLevel != maxLevel || file.deleteRowCount().map(d -> d == 0).orElse(false)) {
            CompactResult upgradeResult = rewriter.upgrade(outputLevel, file);
            toUpdate.merge(upgradeResult);
            upgradeFilesNum++;
        } else {
            // files with delete records should not be upgraded directly to max level
            List<List<SortedRun>> candidate = new ArrayList<>();
            candidate.add(new ArrayList<>());
            candidate.get(0).add(SortedRun.fromSingle(file));
            rewriteImpl(candidate, toUpdate);
        }
    }

    private void rewrite(List<List<SortedRun>> candidate, CompactResult toUpdate) throws Exception {
        if (candidate.isEmpty()) {
            return;
        }
        if (candidate.size() == 1) {
            List<SortedRun> section = candidate.get(0);
            if (section.isEmpty()) {
                return;
            } else if (section.size() == 1) {
                for (DataFileMeta file : section.get(0).files()) {
                    upgrade(file, toUpdate);
                }
                candidate.clear();
                return;
            }
        }
        rewriteImpl(candidate, toUpdate);
    }

    private void rewriteImpl(List<List<SortedRun>> candidate, CompactResult toUpdate)
            throws Exception {
        CompactResult rewriteResult = rewriter.rewrite(outputLevel, dropDelete, candidate);
        toUpdate.merge(rewriteResult);
        candidate.clear();
    }
}

File: paimon-python/pypaimon/compact/task/merge_tree_compact_task.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################

from typing import Any, Dict, List, Tuple

from pypaimon.compact.rewriter.merge_tree_compact_rewriter import \
    MergeTreeCompactRewriter
from pypaimon.compact.task.compact_task import CompactTask, register_compact_task
from pypaimon.manifest.schema.data_file_meta import (DataFileMeta, decode_value,
                                                     encode_value)
from pypaimon.read.interval_partition import IntervalPartition
from pypaimon.read.reader.merge_function import \
    create_merge_function_factory
from pypaimon.write.commit_message import CommitMessage


@register_compact_task
class MergeTreeCompactTask(CompactTask):
    """Compact a single (partition, bucket) of a primary-key table.

    Carries the picked CompactUnit's files plus the strategy-decided
    output_level and drop_delete flag. The driver attaches the in-process
    FileStoreTable; Phase 4 will plumb the loader fields for distributed
    execution.
    """

    TYPE = "merge-tree-compact"

    def __init__(
        self,
        partition: Tuple,
        bucket: int,
        files: List[DataFileMeta],
        output_level: int,
        drop_delete: bool,
        table=None,
    ):
        self.partition = tuple(partition)
        self.bucket = bucket
        self.files = list(files)
        self.output_level = output_level
        self.drop_delete = drop_delete
        self._table = table

    def with_table(self, table) -> "MergeTreeCompactTask":
        self._table = table
        return self

    def run(self) -> CommitMessage:
        table = self._resolve_table()

        # IntervalPartition reproduces split_read.MergeFileSplitRead.create_reader's
        # section grouping so the rewriter sees the same "non-overlapping
        # SortedRuns per section" layout it would on a normal scan.
        sections = IntervalPartition(self.files).partition()

        rewriter = MergeTreeCompactRewriter(
            table=table,
            mf_factory=create_merge_function_factory(table.options),
        )
        after = rewriter.rewrite(
            partition=self.partition,
            bucket=self.bucket,
            output_level=self.output_level,
            sections=sections,
            drop_delete=self.drop_delete,
        )

        return CommitMessage(
            partition=self.partition,
            bucket=self.bucket,
            compact_before=list(self.files),
            compact_after=list(after),
        )

    def _to_payload(self) -> Dict[str, Any]:
        return {
            "partition": [encode_value(v) for v in self.partition],
            "bucket": self.bucket,
            "files": [f.to_dict() for f in self.files],
            "output_level": self.output_level,
            "drop_delete": self.drop_delete,
        }

    @classmethod
    def _from_payload(cls, payload: Dict[str, Any]) -> "MergeTreeCompactTask":
        return cls(
            partition=tuple(decode_value(v) for v in payload.get("partition") or []),
            bucket=payload["bucket"],
            files=[DataFileMeta.from_dict(f) for f in payload.get("files") or []],
            output_level=payload["output_level"],
            drop_delete=payload["drop_delete"],
        )

    def _resolve_table(self):
        if self._table is not None:
            return self._table
        return self._resolve_table_via_loader()
